File: games/js/asteroid.js
/*
 * Misho Georgiev game project
 *
 * 2022 Softel Labs
 *
*/


class Asteroid extends BaseMoveConstantVelocity {
  constructor(ctx, w, h, x0, y0, vx0, vy0, r0, fps_ratio=1, color="white", size_contour=20) {
    super(ctx, w, h, x0, y0, vx0, vy0, fps_ratio, color)

    this.r0 = r0
    this.r = r0
    this.size_contour = size_contour
    this.ox = r0
    this.oy = r0

    this.generate_contour()
  }

  draw() {
    // this.generate_contour()
    let current_list_points = []
    for (let i = 0; i < this.list_points.length; i++) {
      let x1 = this.x + this.list_points[i][0]
      let y1 = this.y + this.list_points[i][1]
      let p1 = [x1, y1]
      current_list_points.push([x1, y1])
    }

    draw_contours(this.ctx, current_list_points, this.color, 1)
    draw_circle(this.ctx, this.x, this.y, this.r, "#AA0000")
  }

  generate_contour() {
    this.list_points = []
    for (let i = 0; i < this.size_contour; i++) {
      let r = this.r0 * Math.sqrt(randuniform(0.3, 1.0, 5))
      let angle = i * 2 * Math.PI / this.size_contour
      let p = [r * Math.cos(angle), r * Math.sin(angle)]
      this.list_points.push(p)
    }
  }
}

File: games/py/asteroid.py
import math
import cv2
import numpy as np

from geometrics import *
from base_move import *


class Asteroid(BaseMoveConstantVelocity):
  def __init__(self, x0, y0, vx0, vy0, r0, color=(255, 0, 0), size_contour=20):
    super().__init__(x0, y0, vx0, vy0, color)

    self.r0 = r0
    self.r = r0
    self.size_contour = size_contour

    self.generate_contour()

  def draw(self, frame):
    # self.generate_contour()
    list_points = []
    for p in self.list_points:
      list_points.append((int(self.x + p[0]), int(self.y + p[1])))
    cv2.drawContours(frame, [np.array(list_points)], 0, self.color, 1)
    cv2.circle(frame, (int(self.x), int(self.y)), int(self.r), (0, 0, 200), thickness=1)

  def generate_contour(self):
    self.list_points = []
    for i in range(self.size_contour):
      r = self.r0 * np.sqrt(np.random.uniform(0.3, 1.0))
      # angle = np.random.uniform() * 2 * PI
      angle = i * 2 * PI / self.size_contour
      p = (r * math.cos(angle), r * math.sin(angle))
      self.list_points.append(p)
